drivers/flash/flash_mspi_nor_mx.c

```diff
@@ -0,0 +1,989 @@
+			ret = flash_mspi_nor_mx_write_enable(flash);
+			if (ret) {
+				goto erase_end;
+			}
+			ret = flash_mspi_nor_mx_unprotect_sector(flash, offset);
+			if (ret) {
+				goto erase_end;
+			}
+			ret = flash_mspi_nor_mx_write_enable(flash);
+			if (ret) {
+				goto erase_end;
+			}
+			ret = flash_mspi_nor_mx_erase_block(flash, offset);
+			if (ret) {
+				goto erase_end;
+			}
+
+			ret = flash_mspi_nor_mx_mem_ready(flash);
+			if (ret) {
+				goto erase_end;
+			}
+
+			offset += SPI_NOR_BLOCK_SIZE;
+		}
+	} else {
+		for (i = 0; i < num_sectors; i++) {
+			/* Check if write enable on each sector or for the whole device */
+			ret = flash_mspi_nor_mx_write_enable(flash);
+			if (ret) {
+				goto erase_end;
+			}
+			ret = flash_mspi_nor_mx_unprotect_sector(flash, offset);
+			if (ret) {
+				goto erase_end;
+			}
+			/* Check if write enable on each sector or for the whole device */
+			ret = flash_mspi_nor_mx_write_enable(flash);
+			if (ret) {
+				goto erase_end;
+			}
+
+			ret = flash_mspi_nor_mx_erase_sector(flash, offset);
+			if (ret) {
+				goto erase_end;
+			}
+
+			ret = flash_mspi_nor_mx_mem_ready(flash);
+			if (ret) {
+				goto erase_end;
+			}
+
+			offset += SPI_NOR_SECTOR_SIZE;
+		}
+	}
+
+erase_end:
+	release(flash);
+
+	return ret;
+}
+
+static const struct flash_parameters *flash_mspi_nor_mx_get_parameters(const struct device *flash)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+
+	return &cfg->flash_param;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void flash_mspi_nor_mx_pages_layout(const struct device *flash,
+					   const struct flash_pages_layout **layout,
+					   size_t *layout_size)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+
+	*layout = &cfg->page_layout;
+	*layout_size = 1;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+static int flash_mspi_nor_mx_init(const struct device *flash)
+{
+	const struct flash_mspi_nor_mx_config *cfg = flash->config;
+	struct flash_mspi_nor_mx_data *data = flash->data;
+	uint8_t vendor_id;
+
+	if (!device_is_ready(cfg->bus)) {
+		LOG_ERR("Controller device is not ready");
+		return -ENODEV;
+	}
+
+	switch (cfg->tar_dev_cfg.io_mode) {
+	case MSPI_IO_MODE_SINGLE:
+	case MSPI_IO_MODE_QUAD:
+	case MSPI_IO_MODE_OCTAL:
+		break;
+	default:
+		LOG_ERR("MSPI bus mode %d not supported", cfg->tar_dev_cfg.io_mode);
+		return -EIO;
+	}
+
+	switch (cfg->tar_dev_cfg.data_rate) {
+	case MSPI_DATA_RATE_SINGLE:
+	case MSPI_DATA_RATE_DUAL:
+		break;
+	default:
+		LOG_ERR("MSPI bus data rate %d not supported", cfg->tar_dev_cfg.data_rate);
+		return -EIO;
+	}
+
+	/* The SPI/DTR is not a valid config of data_mode/data_rate according to the DTS */
+	if ((cfg->tar_dev_cfg.io_mode == MSPI_IO_MODE_SINGLE) &&
+	    (cfg->tar_dev_cfg.data_rate == MSPI_DATA_RATE_DUAL)) {
+		LOG_ERR("MSPI data rate SPI/DTR is not valid");
+		return -EIO;
+	}
+
+	/* At this time only set the io_mode and data rate */
```

Comment:
You can't assume that only setting IO mode and data rate is enough.
Different controller have different defaults when it comes to the member of struct mspi_dev_cfg. Even the initialization process may needed different values. So they should specified and configure clearly.
Thus this should be a CONFIG_ALL.